Sources/FoundationEssentials/String/String+Encoding+Names.swift

```diff
@@ -0,0 +1,265 @@
+            return lhs.scalar.value - 0x20 == rhs.scalar.value
+        }
+        return false
+    }
+}
+
+/// A type to tokenize string for `String.Encoding` names.
+internal protocol StringEncodingNameTokenizer: ~Copyable {
+    associatedtype Token: Equatable
+    init(name: String)
+    mutating func nextToken() throws -> Token?
+}
+
+extension StringEncodingNameTokenizer where Self: ~Copyable {
+    mutating func hasEqualTokens(with other: consuming Self) throws -> Bool {
+        while let myToken = try self.nextToken() {
+            guard let otherToken = try other.nextToken(),
+                  myToken == otherToken else {
+                return false
+            }
+        }
+        return try other.nextToken() == nil
+    }
+}
+
+
+/// A parser that tokenizes a string into `ASCIICaseInsensitiveUnicodeScalar`s.
+private struct ASCIICaseInsensitiveTokenizer: StringEncodingNameTokenizer, ~Copyable {
+    typealias Token = ASCIICaseInsensitiveUnicodeScalar
+
+    enum Error: Swift.Error {
+        case nonASCII
+    }
+
+    let scalars: Substring.UnicodeScalarView
+
+    var _currentIndex: Substring.UnicodeScalarView.Index
+
+    init(name: String) {
+        self.scalars = name._trimmed
+        self._currentIndex = scalars.startIndex
+    }
+
+    mutating func nextToken() throws -> Token? {
+        guard _currentIndex < scalars.endIndex else {
+            return nil
+        }
+        let scalar = scalars[_currentIndex]
+        guard scalar.isASCII else { throw Error.nonASCII }
+        defer {
+            scalars.formIndex(after: &_currentIndex)
+        }
+        return  ASCIICaseInsensitiveUnicodeScalar(scalar)
+    }
+}
+
+
+private extension String {
+    func isEqual<T>(
+        to other: String,
+        tokenizedBy tokenizer: T.Type
+    ) -> Bool where T: StringEncodingNameTokenizer, T: ~Copyable {
+        do {
+            var myTokenizer = T(name: self)
+            let otherTokenizer = T(name: other)
+            return try myTokenizer.hasEqualTokens(with: otherTokenizer)
+        } catch {
+            // Any errors imply that `self` or `other` contains invalid characters.
+            return false
+        }
+    }
+}
+
+
+// MARK: - IANA Charset Names
+
+/// Info about IANA Charset.
+internal struct IANACharset {
+    /// Preferred MIME Name
+    let preferredMIMEName: String?
+
+    /// The name of this charset
+    let name: String
+
+    /// The aliases of this charset
+    let aliases: Array<String>
+
+    var representativeName: String {
+        return preferredMIMEName ?? name
+    }
+
+    init(preferredMIMEName: String?, name: String, aliases: Array<String>) {
+        self.preferredMIMEName = preferredMIMEName
+        self.name = name
+        self.aliases = aliases
+    }
+
+    func matches<T>(
+        _ string: String,
+        tokenizedBy tokenizer: T.Type
+    ) -> Bool where T: StringEncodingNameTokenizer, T: ~Copyable {
+        if let preferredMIMEName = self.preferredMIMEName,
```

Comment:
I'm not sure I understand why we need a tokenizer. This function looks like it is effectively the same as doing case-insensitive-compare on two strings. In the case of ascii strings, it is also effectively the same as calling `lowercased()` and compare. Is my understanding correct? Did you add a tokenizer to avoid extra allocation from calling `lowercased()`? Are there other things that I'm missing?